<!doctype html>
<html class="no-js" lang="en" data-content_root="./">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" /><link rel="next" title="Using GRNPipeline" href="GRN_Example.html" /><link rel="prev" title="Sample (Patient) Level Analysis" href="samplepipeline.html" />

    <!-- Generated with Sphinx 7.2.6 and Furo 2024.01.29 -->
        <title>Using CellPipeline - genereporter 0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=362ab14a" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?v=135e06be" />
    <link rel="stylesheet" type="text/css" href="_static/nbsphinx-code-cells.css?v=2aa19091" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?v=36a5483c" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  --color-brand-primary: #7C4DFF;
  --color-brand-content: #7C4DFF;
  --color-problematic: #7C4DFF;
  --color-api-background: #efeff4;
  --color-api-background-hover: #d4d4d9;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  --color-brand-primary: #7C4DFF;
  --color-brand-content: #7C4DFF;
  --color-problematic: #7C4DFF;
  --color-api-background: #1e2124;
  --color-api-background-hover: #45454a;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  --color-brand-primary: #7C4DFF;
  --color-brand-content: #7C4DFF;
  --color-problematic: #7C4DFF;
  --color-api-background: #1e2124;
  --color-api-background-hover: #45454a;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">genereporter 0.1 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  
  <span class="sidebar-brand-text">genereporter 0.1 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="cellpipeline.html">Cell Type Specific Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="grnpipeline.html">Gene Regulatory Network Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="samplepipeline.html">Sample (Patient) Level Analysis</a></li>
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">Using CellPipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="GRN_Example.html">Using GRNPipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="Sample_Example.html">Using SamplePipeline</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section id="Using-CellPipeline">
<h1>Using CellPipeline<a class="headerlink" href="#Using-CellPipeline" title="Link to this heading">#</a></h1>
<section id="A-cell-type-specific-analysis-and-visualization-tool-for-the-gene-of-interest">
<h2>A cell type specific analysis and visualization tool for the gene of interest<a class="headerlink" href="#A-cell-type-specific-analysis-and-visualization-tool-for-the-gene-of-interest" title="Link to this heading">#</a></h2>
<body><p>This notebook is built to be run automatically, you can just “Run All” cells. Beware: this requires some patience and high computational resources at the moment.</p>
</body><p>First, the data and package are loaded. This may take a minute. Set your gene of interest (GOI) here!</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">sys</span>
<span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;/lustre/groups/ml01/workspace/samantha.bening/Bachelor/&#39;</span><span class="p">)</span>
<span class="kn">from</span> <span class="nn">importlib</span> <span class="kn">import</span> <span class="n">reload</span>
<span class="kn">import</span> <span class="nn">genereporter.cell_pipeline</span> <span class="k">as</span> <span class="nn">cp</span>
<span class="n">reload</span><span class="p">(</span><span class="n">cp</span><span class="p">)</span>

<span class="n">cp</span> <span class="o">=</span> <span class="n">cp</span><span class="o">.</span><span class="n">CellPipeline</span><span class="p">(</span><span class="s2">&quot;/lustre/groups/ml01/workspace/samantha.bening/Bachelor/&quot;</span><span class="p">,</span> <span class="s2">&quot;data2/veo_ibd_balanced.h5ad&quot;</span><span class="p">)</span>

<span class="c1"># set your gene of interest</span>
<span class="n">GOI</span> <span class="o">=</span> <span class="s2">&quot;CASP8&quot;</span>
<span class="c1"># set your cell type of interest</span>
<span class="n">cell_type</span> <span class="o">=</span> <span class="s1">&#39;CD4 T&#39;</span>
</pre></div>
</div>
</div>
<p>Below is a list of possible coarse (level 1) cell types. Choose one of these as your cell type of interest above (cell_type = ‘[your cell type]’) to run the notebook automatically. Of course, you can rerun certain outputs on different cell types as well.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># print cell type names here; easier to select</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Coarse cell types: &quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">cell_type</span> <span class="ow">in</span> <span class="n">cp</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype_l2&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t</span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">cell_type</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Coarse cell types:
        Pericyte
        B
        Endothelial
        CD4 T
        CD8 T
        NK_ILC
        Fibroblast
        Cycling B
        Plasma
        Cycling Myeloid
        Cycling Stroma
        Cycling T
        Epithelial
        Glial
        Myeloid
        Tuft
        Smooth Muscle Cell
        pDC
        Mast
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># UMAP of coarse cell types</span>
<span class="n">cp</span><span class="o">.</span><span class="n">plot_umap</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s2">&quot;celltype_l2&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Cell_Example_5_0.png" src="_images/Cell_Example_5_0.png" />
</div>
</div>
<p>Next, we provide a quick summary of the GOI’s expression class and mean expression level across all cell types.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">expr_sum</span> <span class="o">=</span> <span class="n">cp</span><span class="o">.</span><span class="n">explain_expr_celltypes</span><span class="p">(</span><span class="n">GOI</span><span class="o">=</span><span class="s1">&#39;CASP8&#39;</span><span class="p">)</span>
<span class="n">expr_sum</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Cell type</th>
      <th>Expression class</th>
      <th>Avg. expression over cell type</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>CASP8</th>
      <td>pDC</td>
      <td>low</td>
      <td>0.349</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>CD4 T</td>
      <td>very low</td>
      <td>0.266</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>Cycling T</td>
      <td>very low</td>
      <td>0.262</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>CD8 T</td>
      <td>very low</td>
      <td>0.243</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>NK_ILC</td>
      <td>very low</td>
      <td>0.240</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>Mast</td>
      <td>very low</td>
      <td>0.202</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>B</td>
      <td>very low</td>
      <td>0.140</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>Cycling B</td>
      <td>very low</td>
      <td>0.122</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>Cycling Myeloid</td>
      <td>very low</td>
      <td>0.109</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>Plasma</td>
      <td>very low</td>
      <td>0.109</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>Tuft</td>
      <td>very low</td>
      <td>0.108</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>Myeloid</td>
      <td>very low</td>
      <td>0.104</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>Epithelial</td>
      <td>very low</td>
      <td>0.091</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>Endothelial</td>
      <td>very low</td>
      <td>0.064</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>Cycling Stroma</td>
      <td>very low</td>
      <td>0.056</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>Pericyte</td>
      <td>very low</td>
      <td>0.037</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>Fibroblast</td>
      <td>very low</td>
      <td>0.037</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>Glial</td>
      <td>very low</td>
      <td>0.023</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>Smooth Muscle Cell</td>
      <td>very low</td>
      <td>0.013</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cp</span><span class="o">.</span><span class="n">plot_expressions</span><span class="p">(</span><span class="n">GOI</span><span class="p">,</span> <span class="n">cell_type</span><span class="o">=</span><span class="n">cell_type</span><span class="p">,</span> <span class="n">show_summary</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="c1"># Can change show_summary=False to hide the textual summary of the expression classes (quantile thresholds and cell counts per category)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Cell_Example_8_0.png" src="_images/Cell_Example_8_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Summary for all cells:
Quantile thresholds:
very low: 96.2325, low: 98.8921, middle: 99.4425, high: 99.7479, very high: 99.7500

Number of genes per category:
very_low: 27101
low: 749
middle: 155
high: 86
very_high: 71


Summary for CD4 T cells:
Quantile thresholds:
very low: 96.5912, low: 98.988, middle: 99.4709, high: 99.7479, very high: 99.7500

Number of genes per category:
very_low: 27202
low: 675
middle: 136
high: 78
very_high: 71

</pre></div></div>
</div>
</section>
<section id="Expression-vs.-Detection-visualization">
<h2>Expression vs. Detection visualization<a class="headerlink" href="#Expression-vs.-Detection-visualization" title="Link to this heading">#</a></h2>
<body><p>This can contextualize the expression levels we observe in the standard scanpy plots. In single-cell RNA-seq, only a random sampling of the RNA present in a cell is selected to be sequenced. By pure chance, lowly expressed genes may not be present in all the sampled RNA due to their low prevalance. Here, we can inspect the maximum percentage of expression expected in all genes, specifically our gene of interest.</p>
</body><div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cp</span><span class="o">.</span><span class="n">expression_vs_detection</span><span class="p">(</span><span class="n">GOI</span><span class="p">,</span> <span class="n">cell_type</span><span class="o">=</span><span class="n">cell_type</span><span class="p">)</span>
<span class="c1"># Can add (or remove) &quot;cell_type=cell_type&quot; to plot only the cell type of interest (or across all cell types)</span>
<span class="c1"># todo this section before dotplots etc.</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Cell_Example_10_0.png" src="_images/Cell_Example_10_0.png" />
</div>
</div>
<section id="Automatically-identify-lower-outliers-(clue-to-look-at-celltype-subset)">
<h3>Automatically identify lower outliers (clue to look at celltype subset)<a class="headerlink" href="#Automatically-identify-lower-outliers-(clue-to-look-at-celltype-subset)" title="Link to this heading">#</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cp</span><span class="o">.</span><span class="n">plot_outliers</span><span class="p">(</span><span class="n">GOI</span><span class="p">,</span> <span class="n">outlier_threshold</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">cell_type</span><span class="o">=</span><span class="n">cell_type</span><span class="p">)</span>
<span class="c1"># Can add &quot;cell_type=cell_type&quot; to plot only the cell type of interest</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Cell_Example_12_0.png" src="_images/Cell_Example_12_0.png" />
</div>
</div>
<p>This is how the maximum threshold curve approximation is calculated. This is primarily interesting for our fundamental understanding of the curve’s approximation through the spline’s 3rd derivative’s change points and the linear approximation of this curve.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cp</span><span class="o">.</span><span class="n">fit_spline</span><span class="p">(</span><span class="n">plot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cell_type</span><span class="o">=</span><span class="n">cell_type</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Cell_Example_14_0.png" src="_images/Cell_Example_14_0.png" />
</div>
</div>
<p>These are the top 5 number of outliers, sorted by their distance away from the maximum curve. You can show more or less by changing the head=n parameter.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cp</span><span class="o">.</span><span class="n">list_outliers</span><span class="p">(</span><span class="n">cell_type</span><span class="o">=</span><span class="n">cell_type</span><span class="p">)</span>

<span class="c1"># can show top n number of genes by adding &quot;head=n&quot;</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>log1p(means)</th>
      <th>percent_detected</th>
      <th>distance</th>
      <th>is_outlier</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>HSPA1A</th>
      <td>1.041783</td>
      <td>0.459530</td>
      <td>0.394890</td>
      <td>True</td>
    </tr>
    <tr>
      <th>HSPA1B</th>
      <td>0.911569</td>
      <td>0.453423</td>
      <td>0.339097</td>
      <td>True</td>
    </tr>
    <tr>
      <th>IGKC</th>
      <td>0.581478</td>
      <td>0.103288</td>
      <td>0.320663</td>
      <td>True</td>
    </tr>
    <tr>
      <th>KLF2</th>
      <td>0.865623</td>
      <td>0.474351</td>
      <td>0.299710</td>
      <td>True</td>
    </tr>
    <tr>
      <th>CCL4</th>
      <td>0.509195</td>
      <td>0.062818</td>
      <td>0.299202</td>
      <td>True</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</section>
</section>
<section id="GOI-expression-across-cell-types">
<h2>GOI expression across cell types<a class="headerlink" href="#GOI-expression-across-cell-types" title="Link to this heading">#</a></h2>
<body><p>Now we show the standard scanpy plots of our GOI’s expression across both coarse cell types and fine cell types. The fine cell type automatically shown in the one you set at the beginning of this notebook. You can rerun the cell with other cell types of interest by setting the cell_type=[‘your cell type’] parameter.</p>
</body><div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># GOI expression across coarse cell types</span>
<span class="n">cp</span><span class="o">.</span><span class="n">dotplot</span><span class="p">(</span><span class="n">GOI</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Cell_Example_18_0.png" src="_images/Cell_Example_18_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># GOI expression in fine cell type</span>
<span class="n">cp</span><span class="o">.</span><span class="n">dotplot</span><span class="p">(</span><span class="n">GOI</span><span class="p">,</span> <span class="n">cell_type</span><span class="o">=</span><span class="n">cell_type</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Cell_Example_19_0.png" src="_images/Cell_Example_19_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># GOI expression across coarse cell types</span>
<span class="c1"># This is similar to the coarse cell type dotplot previously, just a different visualization</span>
<span class="n">cp</span><span class="o">.</span><span class="n">matrixplot</span><span class="p">(</span><span class="n">GOI</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Cell_Example_20_0.png" src="_images/Cell_Example_20_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># GOI expression across coarse cell types</span>
<span class="c1"># Individual vertical &quot;lines&quot; correspond to individual cells</span>
<span class="c1"># A more fine grained visual than the mean expression plots shown before</span>
<span class="n">cp</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">GOI</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Cell_Example_21_0.png" src="_images/Cell_Example_21_0.png" />
</div>
</div>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="GRN_Example.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Using GRNPipeline</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="samplepipeline.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Sample (Patient) Level Analysis</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2024, Samantha Bening
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Using CellPipeline</a><ul>
<li><a class="reference internal" href="#A-cell-type-specific-analysis-and-visualization-tool-for-the-gene-of-interest">A cell type specific analysis and visualization tool for the gene of interest</a></li>
<li><a class="reference internal" href="#Expression-vs.-Detection-visualization">Expression vs. Detection visualization</a><ul>
<li><a class="reference internal" href="#Automatically-identify-lower-outliers-(clue-to-look-at-celltype-subset)">Automatically identify lower outliers (clue to look at celltype subset)</a></li>
</ul>
</li>
<li><a class="reference internal" href="#GOI-expression-across-cell-types">GOI expression across cell types</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="_static/documentation_options.js?v=2709fde1"></script>
    <script src="_static/doctools.js?v=888ff710"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/scripts/furo.js?v=32e29ea5"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    </body>
</html>