<!doctype html>
<html class="no-js" lang="en" data-content_root="./">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" /><link rel="next" title="Using GRNPipeline" href="GRN_Example.html" /><link rel="prev" title="Gene Regulatory Network Analysis" href="grnpipeline.html" />

    <!-- Generated with Sphinx 7.2.6 and Furo 2024.01.29 -->
        <title>Using CellPipeline - genereporter 0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=362ab14a" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?v=135e06be" />
    <link rel="stylesheet" type="text/css" href="_static/nbsphinx-code-cells.css?v=2aa19091" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?v=36a5483c" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  --color-brand-primary: #7C4DFF;
  --color-brand-content: #7C4DFF;
  --color-problematic: #7C4DFF;
  --color-api-background: #efeff4;
  --color-api-background-hover: #d4d4d9;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  --color-brand-primary: #7C4DFF;
  --color-brand-content: #7C4DFF;
  --color-problematic: #7C4DFF;
  --color-api-background: #1e2124;
  --color-api-background-hover: #45454a;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  --color-brand-primary: #7C4DFF;
  --color-brand-content: #7C4DFF;
  --color-problematic: #7C4DFF;
  --color-api-background: #1e2124;
  --color-api-background-hover: #45454a;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">genereporter 0.1 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  
  <span class="sidebar-brand-text">genereporter 0.1 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="cellpipeline.html">Cell Type Specific Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="grnpipeline.html">Gene Regulatory Network Analysis</a></li>
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">Using CellPipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="GRN_Example.html">Using GRNPipeline</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section id="Using-CellPipeline">
<h1>Using CellPipeline<a class="headerlink" href="#Using-CellPipeline" title="Link to this heading">#</a></h1>
<section id="A-cell-type-specific-analysis-and-visualization-tool-for-the-gene-of-interest">
<h2>A cell type specific analysis and visualization tool for the gene of interest<a class="headerlink" href="#A-cell-type-specific-analysis-and-visualization-tool-for-the-gene-of-interest" title="Link to this heading">#</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">gene_report.cell_pipeline</span> <span class="kn">import</span> <span class="n">CellPipeline</span>

<span class="n">cp</span> <span class="o">=</span> <span class="n">CellPipeline</span><span class="p">(</span><span class="s2">&quot;/Users/samibening/Projects/Bachelor/&quot;</span><span class="p">,</span> <span class="s2">&quot;data/output/adata.h5ad&quot;</span><span class="p">)</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">cp</span><span class="o">.</span><span class="n">get_adata</span><span class="p">()</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 5397 × 16719
    obs: &#39;sampleID&#39;, &#39;barcode&#39;, &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_20_genes&#39;, &#39;total_counts_mt&#39;, &#39;log1p_total_counts_mt&#39;, &#39;pct_counts_mt&#39;, &#39;total_counts_ribo&#39;, &#39;log1p_total_counts_ribo&#39;, &#39;pct_counts_ribo&#39;, &#39;total_counts_hb&#39;, &#39;log1p_total_counts_hb&#39;, &#39;pct_counts_hb&#39;, &#39;outlier&#39;, &#39;mt_outlier&#39;, &#39;_scvi_batch&#39;, &#39;_scvi_labels&#39;, &#39;leiden_res0_6&#39;, &#39;manual_celltype_annotation&#39;, &#39;celltypist_cell_label&#39;, &#39;celltypist_conf_score&#39;, &#39;celltypist_cell_label_coarse&#39;
    var: &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;_scvi_manager_uuid&#39;, &#39;_scvi_uuid&#39;, &#39;celltypist_cell_label_coarse_colors&#39;, &#39;celltypist_cell_label_colors&#39;, &#39;hvg&#39;, &#39;leiden&#39;, &#39;leiden_res0_6_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;sampleID_colors&#39;, &#39;umap&#39;
    obsm: &#39;X_pca&#39;, &#39;X_scVI&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;int_norm&#39;, &#39;log_int_norm&#39;, &#39;log_norm&#39;, &#39;norm&#39;, &#39;raw&#39;
    obsp: &#39;connectivities&#39;, &#39;distances&#39;
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># TODO: print cell type names here? Easier to select?</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># set your gene of interest (GOI)</span>
<span class="n">GOI</span> <span class="o">=</span> <span class="s2">&quot;CASP8&quot;</span>

<span class="c1"># set your cell type of interest (cell_type)</span>
<span class="c1"># set this later on, after expression across cell type plots? Then can&#39;t run the entire notebook easily</span>
<span class="n">cell_type</span> <span class="o">=</span> <span class="s1">&#39;T Cell&#39;</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># UMAP of coarse cell types</span>
<span class="n">cp</span><span class="o">.</span><span class="n">plot_umap</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Cell_Example_4_0.png" src="_images/Cell_Example_4_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">expr_sum</span> <span class="o">=</span> <span class="n">cp</span><span class="o">.</span><span class="n">explain_expr_celltypes</span><span class="p">(</span><span class="n">GOI</span><span class="p">)</span>
<span class="n">expr_sum</span>
<span class="c1"># TODO: sort this by expression high to low?</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Cell type</th>
      <th>Expression class</th>
      <th>Avg. expression over cell type</th>
    </tr>
    <tr>
      <th>index</th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>CASP8</th>
      <td>mLTo</td>
      <td>very low</td>
      <td>0.034</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>Monocytes</td>
      <td>very low</td>
      <td>0.171</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>B Cell</td>
      <td>very low</td>
      <td>0.221</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>Epithelial</td>
      <td>very low</td>
      <td>0.135</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>T Cell</td>
      <td>low</td>
      <td>0.399</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>Dentritic</td>
      <td>very low</td>
      <td>0.162</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>Plasma</td>
      <td>very low</td>
      <td>0.184</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>Enteroendocrine</td>
      <td>very low</td>
      <td>0.163</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>Colonocyte</td>
      <td>very low</td>
      <td>0.120</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>myofibroblast</td>
      <td>very low</td>
      <td>0.141</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>Stromal</td>
      <td>very low</td>
      <td>0.120</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>Adult Glia</td>
      <td>very low</td>
      <td>0.027</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>Pericyte</td>
      <td>very low</td>
      <td>0.054</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>ILC3</td>
      <td>very low</td>
      <td>0.266</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>NK cell</td>
      <td>very low</td>
      <td>0.357</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>Macrophages</td>
      <td>very low</td>
      <td>0.269</td>
    </tr>
    <tr>
      <th>CASP8</th>
      <td>Goblet</td>
      <td>very low</td>
      <td>0.246</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cp</span><span class="o">.</span><span class="n">plot_expressions</span><span class="p">(</span><span class="n">GOI</span><span class="p">,</span> <span class="n">cell_type</span><span class="p">,</span> <span class="n">show_summary</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="c1"># Can change show_summary=True to show the summary of the expression classes (quantile thresholds and cell counts per category)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Cell_Example_6_0.png" src="_images/Cell_Example_6_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Summary for all cells:
Quantile thresholds:
very low: 90.8188, low: 97.769, middle: 99.0909, high: 99.7488, very high: 99.7500

Number of genes per category:
very_low: 15184
low: 1162
middle: 221
high: 110
very_high: 42


Summary for T Cell cells:
Quantile thresholds:
very low: 90.2805, low: 97.5298, middle: 99.1208, high: 99.7488, very high: 99.7500

Number of genes per category:
very_low: 15094
low: 1212
middle: 266
high: 105
very_high: 42

</pre></div></div>
</div>
</section>
<section id="GOI-expression-across-cell-types">
<h2>GOI expression across cell types<a class="headerlink" href="#GOI-expression-across-cell-types" title="Link to this heading">#</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># find expression across cell types of GOI here</span>
<span class="n">cp</span><span class="o">.</span><span class="n">dotplot</span><span class="p">(</span><span class="n">GOI</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Cell_Example_8_0.png" src="_images/Cell_Example_8_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cp</span><span class="o">.</span><span class="n">dotplot</span><span class="p">(</span><span class="n">GOI</span><span class="p">,</span> <span class="n">cell_type</span><span class="o">=</span><span class="n">cell_type</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Cell_Example_9_0.png" src="_images/Cell_Example_9_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cp</span><span class="o">.</span><span class="n">matrixplot</span><span class="p">(</span><span class="n">GOI</span><span class="p">,</span> <span class="n">cell_type</span><span class="o">=</span><span class="n">cell_type</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Cell_Example_10_0.png" src="_images/Cell_Example_10_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cp</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">GOI</span><span class="p">)</span>
<span class="c1"># show different normalization methods here?</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Cell_Example_11_0.png" src="_images/Cell_Example_11_0.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># TODO: patient specific data:</span>
<span class="c1"># Any specific things here?</span>
</pre></div>
</div>
</div>
</section>
<section id="Expression-vs.-Detection-visualization">
<h2>Expression vs. Detection visualization<a class="headerlink" href="#Expression-vs.-Detection-visualization" title="Link to this heading">#</a></h2>
<body><p>This can contextualize the expression levels we observed previously. In single-cell RNA-seq, only a random sampling of the RNA present in a cell is selected to be sequenced. By pure chance, lowly expressed genes may not be present in all the sampled RNA due to their low prevalance. Here, we can inspect the maximum percentage of expression expected in all genes, specifically our gene of interest.</p>
</body><div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cp</span><span class="o">.</span><span class="n">expression_vs_detection</span><span class="p">(</span><span class="n">GOI</span><span class="p">,</span> <span class="n">cell_type</span><span class="o">=</span><span class="n">cell_type</span><span class="p">)</span>
<span class="c1"># Can add &quot;cell_type=cell_type&quot; to plot only the cell type of interest</span>
<span class="c1"># rather have a side by side view here again?</span>
<span class="c1"># todo this section before dotplots etc.</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Cell_Example_14_0.png" src="_images/Cell_Example_14_0.png" />
</div>
</div>
<section id="Automatically-identify-lower-outliers-(clue-to-look-at-celltype-subset)">
<h3>Automatically identify lower outliers (clue to look at celltype subset)<a class="headerlink" href="#Automatically-identify-lower-outliers-(clue-to-look-at-celltype-subset)" title="Link to this heading">#</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cp</span><span class="o">.</span><span class="n">plot_outliers</span><span class="p">(</span><span class="n">GOI</span><span class="p">,</span> <span class="n">cell_type</span><span class="o">=</span><span class="n">cell_type</span><span class="p">)</span>
<span class="c1"># Can add &quot;cell_type=cell_type&quot; to plot only the cell type of interest</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Cell_Example_16_0.png" src="_images/Cell_Example_16_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cp</span><span class="o">.</span><span class="n">list_outliers</span><span class="p">()</span>
<span class="c1"># show top 5 most extreme outlier genes</span>
<span class="c1"># (by orthogonal distance to the max line)</span>
<span class="c1"># can show top x genes by adding &quot;head=x&quot; to the function</span>

<span class="c1">#TODO: make outlier plot above interactive? Hover over dot to show gene name etc. (sami: use plotly)</span>
<span class="c1">#      Slider for threshold value?</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>log1p(means)</th>
      <th>percent_detected</th>
      <th>distance</th>
      <th>is_outlier</th>
    </tr>
    <tr>
      <th>index</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>CXCL14</th>
      <td>1.002802</td>
      <td>0.075783</td>
      <td>0.848793</td>
      <td>True</td>
    </tr>
    <tr>
      <th>IGLC7</th>
      <td>0.932603</td>
      <td>0.045951</td>
      <td>0.772127</td>
      <td>True</td>
    </tr>
    <tr>
      <th>IGHG1</th>
      <td>0.963830</td>
      <td>0.067074</td>
      <td>0.766993</td>
      <td>True</td>
    </tr>
    <tr>
      <th>IGHM</th>
      <td>1.242698</td>
      <td>0.212525</td>
      <td>0.760515</td>
      <td>True</td>
    </tr>
    <tr>
      <th>IGFBP7</th>
      <td>1.167887</td>
      <td>0.246063</td>
      <td>0.713274</td>
      <td>True</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</section>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="GRN_Example.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Using GRNPipeline</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="grnpipeline.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Gene Regulatory Network Analysis</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2024, sami
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Using CellPipeline</a><ul>
<li><a class="reference internal" href="#A-cell-type-specific-analysis-and-visualization-tool-for-the-gene-of-interest">A cell type specific analysis and visualization tool for the gene of interest</a></li>
<li><a class="reference internal" href="#GOI-expression-across-cell-types">GOI expression across cell types</a></li>
<li><a class="reference internal" href="#Expression-vs.-Detection-visualization">Expression vs. Detection visualization</a><ul>
<li><a class="reference internal" href="#Automatically-identify-lower-outliers-(clue-to-look-at-celltype-subset)">Automatically identify lower outliers (clue to look at celltype subset)</a></li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="_static/documentation_options.js?v=2709fde1"></script>
    <script src="_static/doctools.js?v=888ff710"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/scripts/furo.js?v=32e29ea5"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    </body>
</html>